config/src/main/java/com/typesafe/config/impl/BeanFactory.java

```diff
@@ -0,0 +1,160 @@
+                if (configValue == null) {
+                    throw new ConfigException.Generic(
+                            "Could not find " + beanProp.getName() + " from " + clazz.getName() + " in config.");
+                }
+                if (configValue instanceof Map) {
+                    configValue = createInternal(config.getConfig(originalNames.get(beanProp.getDisplayName())), beanProp.getPropertyType());
+                } else {
+                    Class parameterClass = setter.getParameterTypes()[0];
+                    configValue = getValueWithAutoConversion(parameterClass, config, originalNames.get(beanProp.getDisplayName()));
+                }
+                setter.invoke(bean, configValue);
+
+            }
+            return bean;
+        } catch (IntrospectionException e) {
+            throw new ConfigException.Generic("Could not resolve a string method name.", e);
+        } catch (InstantiationException e) {
+            throw new ConfigException.Generic(clazz + " needs a public no args constructor", e);
+        } catch (IllegalAccessException e) {
+            throw new ConfigException.Generic(clazz + " getters and setters are not accessible", e);
+        } catch (InvocationTargetException e) {
+            throw new ConfigException.Generic("Calling bean method caused exception", e);
+        }
+    }
+
+    private static Object getValueWithAutoConversion(Class parameterClass, Config config, String configPropName) {
+        if (parameterClass == Boolean.class || parameterClass == boolean.class) {
+            return config.getBoolean(configPropName);
+        } else if (parameterClass == Integer.class || parameterClass == int.class) {
+            return config.getInt(configPropName);
+        } else if (parameterClass == Double.class || parameterClass == double.class) {
+            return config.getDouble(configPropName);
+        } else if (parameterClass == Long.class || parameterClass == long.class) {
+            String rawVal = config.getString(configPropName);
+            if (isDuration(rawVal)) {
+                TimeUnit timeUnit = getTimeUnit(rawVal);
+                return config.getDuration(configPropName, timeUnit);
+            } else if (isByteValue(rawVal)) {
+                return config.getBytes(configPropName);
+            }
+            return config.getLong(configPropName);
+        } else if (parameterClass == String.class) {
+            return config.getString(configPropName);
+        }
+
+        return config.getAnyRef(configPropName);
+    }
+
+    private static boolean isDuration(String rawVal) {
+        for (Map.Entry<TimeUnit, List<String>> timeUnitListEntry : TIME_UNITS.entrySet()) {
+            for (String timeUnitStr : timeUnitListEntry.getValue()) {
+                if(rawVal.contains(timeUnitStr)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+
+    static TimeUnit getTimeUnit(String rawVal) {
```

Comment:
My idea of not using parseDuration was that in case of BeanConfig I try to perform autoconversion of duration (and bytes too) only if marker representing time (size) unit (i.e. 's','ms','seconds' etc) is present in rawString. I have no meta information how to treat string --> long conversion so I inspect string for some 'markers' for duration or size and then perform conversion. Of course if we can have come meta-info annotations then it will be possible to reuse parseDuratation. Another way may be to override parseDuration with some flag indicating that we should parse only string containing markers.